include/algorithms/synthesis/encoding.hpp

```diff
@@ -0,0 +1,159 @@
+#ifndef ENCODING_HPP
+#define ENCODING_HPP
+
+#include "core/truthTable/truth_table.hpp"
+
+namespace syrec {
+
+    struct MinHeapNode {
+        TruthTable::Cube data;
+
+        std::size_t freq{};
+
+        std::shared_ptr<MinHeapNode> left{};
+        std::shared_ptr<MinHeapNode> right{};
+
+        MinHeapNode(TruthTable::Cube data, const std::size_t freq):
+            data(std::move(data)), freq(freq) {}
+
+        auto operator>(const MinHeapNode& other) const -> bool {
+            return freq > other.freq;
+        }
+
+        auto traverse(TruthTable::Cube&& encodedCube, TruthTable::CubeMap& encoding) const -> void {
+            // leaf node -> add encoding
+            if (!data.empty()) {
+                encoding.try_emplace(data, std::move(encodedCube));
+                return;
+            }
+
+            // non-leaf node -> traverse left and right subtree
+            if (left) {
+                left->traverse(encodedCube.appendZero(), encoding);
+            }
+            if (right) {
+                right->traverse(encodedCube.appendOne(), encoding);
+            }
+        }
+    };
+
+    inline auto extend(TruthTable& tt) -> void {
+        // ensure that the resulting complete table can be stored in the cube map (at most 63 inputs, probably less in practice)
+        assert(tt.cubeMap.empty() || (tt.nInputs() <= std::log2(tt.cubeMap.max_size()) && tt.nInputs() <= 63U));
```

Comment:
this assert only triggers for debug configurations. however, you want to make sure that an overflow can never happen.
convert this to proper checks outside of an `assert` and throw an exception if the conditions are not satisfied.

Reply:
done 👍🏼